### PR TITLE
Manage  NetworkLink tags in KML file

### LINKED ISSUE #1918: Support NetworkLink in KML Import

### PR DESCRIPTION
**Pull request by oterral**
Fix  #1918 

This PR  uses promises in gaKml in order to get features from NetworkLink tags

[Test](http://mf-geoadmin3.dev.bgdi.ch/teo_net_links/index.html?X=166266.05&Y=633389.01&zoom=2&lang=de&topic=ech&bgLayer=ch.swisstopo.pixelkarte-farbe&layers=KML%7C%7Chttps:%2F%2Fdoc-10-38-docs.googleusercontent.com%2Fdocs%2Fsecuresc%2Flpbfmkon86t1kr9s1hfh5ofs0s235fnq%2Fk4o1pfjltrtf2kbk3jhbapoosgk6drlr%2F1424188800000%2F10258287539236638996%2F10258287539236638996%2F0B4uRuJ1hajZmdld1bzdkVDEzMkszZHFjdHhtSzNDTGdpSkM4%3Fe%3Ddownload%26gd%3Dtrue%26access_token%3Dya29.HQFvAplD39zTDfSlDWbyYNiAXwNCGxrjCy7_Gxd_q2aTeBD8_8cPC4f4Ms6dw5Wc0iTmNQ7zA0ZnWg)

This test link includes a KML  file including 3 others KML files via NetworLink tags

### CODE REVIEW

**Comment by procrastinatio** [comment]
OK, I test with this [KML](http://mf-chsdi3.dev.bgdi.ch/~ltmom/kml/network_link.kml) linking another [one](web-storymaps.dev.bgdi.ch/cgi-bin/kml.cgi) generating a random point.

The `NetworkLink`is refreshed on reload, but not on moving around, which is fine.

Print fails (`Cannot set property '_gx_style' of undefined`)

File: src/components/map/MapService.js [hunk 1]
```diff
@@ -405,19 +405,114 @@
...
+      // Read a kml string then return a list of features.
+      var readFeatures = function(kml) {
+        // Replace all hrefs to prevent errors if image doesn't have
+        // CORS headers. Exception for google markers icons (lightblue.png,
+        // ltblue-dot.png, ltblu-pushpin.png, ...) to keep the OL3 magic for
+        // anchor origin.
+        // Test regex here: http://regex101.com/r/tF3vM0
+        // List of google icons: http://www.lass.it/Web/viewer.aspx?id=4
+        kml = kml.replace(
+          /<href>http(s?)(?!(s?):\/\/maps\.(?:google|gstatic)\.com.*(blue|green|orange|pink|purple|red|yellow|pushpin).*\.png)/g,
```
Comment (procrastinatio) [inline src/components/map/MapService.js]: what about 'brown' or 'white' ?

**Comment by oterral** [comment]
Print is fixed with https://github.com/geoadmin/mf-geoadmin3/pull/2157

**Comment by oterral** [comment]
@ponceta you want to test this ?

**Comment by ponceta** [comment]
I'm already on it ;)

**Comment by oterral** [comment]
@ponceta have you seen bug?

**Comment by ponceta** [comment]
Really nice, works fine on IE 11, chrome and FF.

File: src/components/map/MapService.js [hunk 2]
```diff
@@ -405,19 +405,114 @@
...
-         */
+
+      // Read a kml string then return a list of features.
+      var readFeatures = function(kml) {
+        // Replace all hrefs to prevent errors if image doesn't have
+        // CORS headers. Exception for google markers icons (lightblue.png,
+        // ltblue-dot.png, ltblu-pushpin.png, ...) to keep the OL3 magic for
+        // anchor origin.
+        // Test regex here: http://regex101.com/r/tF3vM0
+        // List of google icons: http://www.lass.it/Web/viewer.aspx?id=4
```
Comment (gjn) [inline src/components/map/MapService.js]: How stable is this list? Would be hard to maintain it if it changes on a regular basis.

Or: why not keep all google png's in there as google is likely to support CORS anyway?

File: src/components/map/MapService.js [hunk 3]
```diff
@@ -405,19 +405,114 @@
...
+          '<href>' + gaGlobalOptions.ogcproxyUrl + 'http'
+        );
+
+        var all = [];
+        var features = kmlFormat.readFeatures(kml);
+        var networkLinks = kmlFormat.readNetworkLinks(kml);
+        if (networkLinks.length) {
+          angular.forEach(networkLinks, function(networkLink) {
+            if (gaUrlUtils.isValid(networkLink.href)) {
+              all.push($http.get(networkLink.href).success(function(data) {
```
Comment (gjn) [inline src/components/map/MapService.js]: What happens on failure?

**Comment by gjn** [comment]
LGTM. Have 2 questions only.

File: src/components/map/MapService.js (hunk 2 above)
Comment (oterral) [inline src/components/map/MapService.js]: It was like that before, it's not came with this PR. 

Google pngs supports CORS, the regex replace all icons href except google icons.

File: src/components/map/MapService.js (hunk 3 above)
Comment (oterral) [inline src/components/map/MapService.js]: Nothing. It's intentional. The user is responsible for setting good URL in his KML.

**Pull request merged**